service/collector/issue.go

```diff
@@ -0,0 +1,157 @@
+package collector
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/google/go-github/github"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var (
+	issueLabelsDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, "labels_count"),
+		"Github issue labels.",
+		[]string{
+			labelOrg,
+			labelRepo,
+			labelLabel,
+			labelState,
+		},
+		nil,
+	)
+	issueStatesDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystem, "states_count"),
+		"Github issue states.",
+		[]string{
+			labelOrg,
+			labelRepo,
+			labelState,
+		},
+		nil,
+	)
+)
+
+type IssueConfig struct {
+	GithubClient *github.Client
+	Logger       micrologger.Logger
+}
+
+type Issue struct {
+	githubClient *github.Client
+	logger       micrologger.Logger
+}
+
+func NewIssue(config IssueConfig) (*Issue, error) {
+	if config.GithubClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.GithubClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	i := &Issue{
+		githubClient: config.GithubClient,
+		logger:       config.Logger,
+	}
+
+	return i, nil
+}
+
+func (i *Issue) Collect(ch chan<- prometheus.Metric) error {
+	ctx := context.Background()
+	opts := &github.IssueListByRepoOptions{
+		ListOptions: github.ListOptions{
+			Page: 1,
+			// The maximum result in the API is 1000, which it not always guarantees.
+			// In manual tests the number of issues received was 100. See also
+			// https://developer.github.com/v3/search/#about-the-search-api.
+			PerPage: 1000,
+		},
+		State: "all",
+	}
+
+	issueLabels := map[string]float64{}
+	issueStates := map[string]float64{}
+	for {
+		issues, res, err := i.githubClient.Issues.ListByRepo(ctx, githubOrg, githubRepo, opts)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		i.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("collecting %3d issues of page %2d", len(issues), opts.Page))
+
+		for _, issue := range issues {
+			if issue.IsPullRequest() {
+				continue
+			}
+
+			for _, label := range issue.Labels {
+				k := fmt.Sprintf("%s:%s", label.GetName(), issue.GetState())
```

Comment:
I think it would be more transparent to use a dedicated type here. Something like:

```go
type issueLabel struct {
	Name  string
	State string
}
```

You can create type `map[issueLabel]float64` then.

See https://play.golang.org/p/4ZUrk_WdKPu

Reply:
Nice. I thought about it but was too lazy to verify because I was not sure the struct key is representative. I will change this.

Reply:
Sorted.